runtime/src/comm/ofi/comm-ofi.c

```diff
@@ -2099,11 +2107,51 @@ void init_ofiForAms(void) {
 }
 
 
-#if 0
+static inline void amRequestNop(c_nodeid_t, chpl_bool, struct perTxCtxInfo_t*);
+
 static
-void init_ofiPerThread(void) {
+void init_ofiConnections(void) {
+  //
+  // With providers that do dynamic endpoint connection we have seen
+  // fairly dramatic connection overheads under certain circumstances.
+  // As an aid to performance analysis, here we allow for forcing the
+  // endpoint connections to be established early, during startup,
+  // rather than later during timed sections of user code.
+  //
+  if (!providerInUse(provType_tcp)
+      && !providerInUse(provType_verbs)) {
+    return;
+  }
+
+  if (!chpl_env_rt_get_bool("COMM_OFI_CONNECT_EAGERLY", false)) {
+    return;
+  }
+
+  //
+  // We do this by firing a no-op AM to every remote node from every
+  // still-inactive tx context.  (In effect, this means from all tx
+  // contexts that aren't already bound to AM handlers.)  At present we
+  // use blocking AMs for this, but if needed we could delay blocking
+  // until after the last one had been initiated.  One way or another,
+  // we need to not return until after all the connections have been
+  // established.
+  //
+  for (int i = 0; i < tciTabLen; i++) {
+    struct perTxCtxInfo_t* tcip = &tciTab[i];
+    if (!tcip->bound) {
+      CHK_TRUE(tciAllocTabEntry(tcip));
+      for (c_nodeid_t node = 0; node < chpl_numNodes; node++) {
+        if (node != chpl_nodeID) {
+          while (tcip->txCQ != NULL && tcip->numTxnsOut >= txCQLen) {
+            (*tcip->checkTxCmplsFn)(tcip);
+          }
+          amRequestNop(node, true /*blocking*/, tcip);
+        }
+      }
+      tciFree(tcip);
```

Comment:
To you need to do `tciFree` here, it looks like it's done in the common am code already (not positive about that though, so should double check me)

Reply:
It's only done in the common code if that common code had to allocate a tx ctx itself, i.e. if the caller didn't supply one. If the caller supplied one it's left allocated.